src/util/secrets/secrets.c

```diff
@@ -0,0 +1,1235 @@
+    TALLOC_CTX *tmp_ctx;
+    errno_t ret;
+
+    if (_sec_ctx == NULL) {
+        return EINVAL;
+    }
+
+    tmp_ctx = talloc_new(mem_ctx);
+    if (tmp_ctx == NULL) {
+        return ENOMEM;
+    }
+
+    sec_ctx = talloc_zero(tmp_ctx, struct sss_sec_ctx);
+    if (sec_ctx == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = set_quotas(sec_ctx, config_list);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_MINOR_FAILURE, "Failed to set quotas\n");
+        /* Not fatal */
+    }
+
+    sec_ctx->ldb = ldb_init(sec_ctx, NULL);
+    if (sec_ctx->ldb == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = ldb_connect(sec_ctx->ldb, dbpath, 0, NULL);
+    if (ret != LDB_SUCCESS) {
+        DEBUG(SSSDBG_TRACE_LIBS,
+              "ldb_connect(%s) returned %d: %s\n",
+              dbpath, ret, ldb_strerror(ret));
+        talloc_free(sec_ctx->ldb);
+        ret = EIO;
+        goto done;
+    }
+
+    ret = lcl_read_mkey(sec_ctx, &sec_ctx->master_key);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Cannot get the master key\n");
+        goto done;
+    }
+
+    ret = EOK;
+done:
+    if (ret == EOK) {
+        *_sec_ctx = talloc_steal(mem_ctx, sec_ctx);
+    }
+    talloc_free(tmp_ctx);
+    return ret;
+}
+
+static int local_db_dn(TALLOC_CTX *mem_ctx,
+                       struct ldb_context *ldb,
+                       const char *basedn,
+                       const char *req_path,
+                       struct ldb_dn **req_dn)
+{
+    struct ldb_dn *dn;
+    const char *s, *e;
+    int ret;
+
+    dn = ldb_dn_new(mem_ctx, ldb, basedn);
+    if (!dn) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    s = req_path;
+
+    while (s && *s) {
+        e = strchr(s, '/');
+        if (e) {
+            if (e == s) {
+                s++;
+                continue;
+            }
+            if (!ldb_dn_add_child_fmt(dn, "cn=%.*s", (int)(e - s), s)) {
+                ret = ENOMEM;
+                goto done;
+            }
+            s = e + 1;
+        } else {
+            if (!ldb_dn_add_child_fmt(dn, "cn=%s", s)) {
+                ret = ENOMEM;
+                goto done;
+            }
+            s = NULL;
+        }
+    }
+
+    DEBUG(SSSDBG_TRACE_INTERNAL,
+          "Local path for [%s] is [%s]\n",
+          req_path, ldb_dn_get_linearized(dn));
+    *req_dn = dn;
+    ret = EOK;
+
+done:
+    return ret;
+}
+
+errno_t sss_sec_new_req(TALLOC_CTX *mem_ctx,
+                        struct sss_sec_ctx *sec_ctx,
+                        const char *url,
+                        uid_t client,
+                        struct sss_sec_req **_req)
+{
+    struct sss_sec_req *req;
+    TALLOC_CTX *tmp_ctx;
+    errno_t ret;
+
+    if (sec_ctx == NULL || url == NULL || _req == NULL) {
+        return EINVAL;
+    }
+
+    tmp_ctx = talloc_new(mem_ctx);
+    if (tmp_ctx == NULL) {
+        return ENOMEM;
+    }
+
+    req = talloc_zero(mem_ctx, struct sss_sec_req);
```

Comment:
Or was your plan to allocate req into tmp_ctx and not mem_ctx?

Reply:
Yes, the current code leaks req on failure. Nice catch.